# TensorFlow. Dense

In [316]:
import tensorflow as tf
import numpy as np
import pandas as pd

from sklearn.datasets import load_boston

from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.models import Model

from sklearn.model_selection import train_test_split

from sklearn.preprocessing import RobustScaler

import matplotlib.pyplot as plt

from tensorflow.keras.optimizers import Adam, SGD


from sklearn.metrics import r2_score

tf.random.set_seed(1)

In [294]:
boston = load_boston()
X = boston.data
y = boston.target

/home/ekaterina/anaconda3/lib/python3.9/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housin

In [295]:
X.shape

(506, 13)

In [296]:
y.shape

(506,)

In [297]:
x_train, x_test, y_train, y_test =train_test_split(X, y, test_size = 0.25, random_state = 100)

In [298]:
scaler = RobustScaler()
scaler = scaler.fit(x_train)

In [299]:
x_train_scaled = scaler.transform(x_train)

In [300]:
x_test_scaled = scaler.transform(x_test)

In [301]:
x_train.shape

(379, 13)

In [302]:
x_test.shape

(127, 13)

In [303]:
y_train[0:5]

array([20. , 20. , 19.7, 22.8, 18.2])

In [304]:
result_list = []

## Создадим 10 различных моделей полносвязной нейросети с различныими параметрами и сравним их по метрике R2. Так как мы решаем задачу регрессии, эта метрика представляется оптимальной. 

## Также в таблицу будем записывать метрики mse и mae.

# Модель 1

In [305]:
model_num = 1

params = '13 input, 1 hidden layer (10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(10, activation='relu')(inputs)
outputs = Dense(1)(hidden_layer_1)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 315.5125 - mae: 14.6729 - val_loss: 231.5181 - val_mae: 11.9464 - 301ms/epoch - 9ms/step
Epoch 2/100
35/35 - 0s - loss: 144.6134 - mae: 9.5324 - val_loss: 107.0374 - val_mae: 7.7727 - 44ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 72.7162 - mae: 6.4490 - val_loss: 70.5455 - val_mae: 6.4477 - 42ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 53.1461 - mae: 5.4733 - val_loss: 54.0655 - val_mae: 5.3576 - 37ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 47.6871 - mae: 5.1095 - val_loss: 97.5497 - val_mae: 8.7428 - 40ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 46.5618 - mae: 5.1798 - val_loss: 48.8821 - val_mae: 5.0877 - 42ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 44.3185 - mae: 4.9358 - val_loss: 45.5418 - val_mae: 4.4159 - 39ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 42.0682 - mae: 4.6701 - val_loss: 50.9842 - val_mae: 5.5232 - 43ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 40.0895 - mae: 4.6106 - val_loss: 56.9734 

Epoch 72/100
35/35 - 0s - loss: 32.1051 - mae: 4.0181 - val_loss: 48.0685 - val_mae: 5.9106 - 39ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 31.6973 - mae: 4.1249 - val_loss: 45.4699 - val_mae: 5.6749 - 39ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 31.5851 - mae: 4.1912 - val_loss: 36.4958 - val_mae: 4.4820 - 38ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 31.5022 - mae: 4.0560 - val_loss: 44.8060 - val_mae: 4.6859 - 40ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 32.1011 - mae: 4.0350 - val_loss: 45.7128 - val_mae: 5.7070 - 36ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 32.0092 - mae: 4.1182 - val_loss: 34.9352 - val_mae: 4.5285 - 37ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 31.5075 - mae: 4.0990 - val_loss: 33.7881 - val_mae: 4.4248 - 41ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 30.8690 - mae: 3.9915 - val_loss: 37.3642 - val_mae: 4.9191 - 39ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 30.9130 - mae: 4.0560 - val_loss: 33.776

'model_num: 1, params: (13 input, 1 hidden layer (10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 36.3, test_mae: 4.29, test_r2: 0.6321'

# Модель 2

In [306]:
model_num = 2

params = '13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='relu')(inputs)
outputs = Dense(1)(hidden_layer_1)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 325.4889 - mae: 13.0496 - val_loss: 102.8406 - val_mae: 7.0683 - 301ms/epoch - 9ms/step
Epoch 2/100
35/35 - 0s - loss: 86.7934 - mae: 6.9170 - val_loss: 113.3347 - val_mae: 9.3253 - 42ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 65.5571 - mae: 6.2673 - val_loss: 123.9983 - val_mae: 10.0905 - 40ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 64.7120 - mae: 6.0770 - val_loss: 82.2329 - val_mae: 6.0646 - 44ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 60.2022 - mae: 5.7315 - val_loss: 346.1761 - val_mae: 17.0254 - 38ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 67.9313 - mae: 6.1455 - val_loss: 48.5713 - val_mae: 4.6502 - 40ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 57.9696 - mae: 5.6445 - val_loss: 161.5051 - val_mae: 10.6885 - 39ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 51.8656 - mae: 5.2947 - val_loss: 76.5927 - val_mae: 6.5706 - 42ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 52.6582 - mae: 5.4755 - val_loss: 124.

Epoch 72/100
35/35 - 0s - loss: 33.5298 - mae: 4.3537 - val_loss: 122.0072 - val_mae: 9.7314 - 39ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 31.8193 - mae: 4.1936 - val_loss: 305.4152 - val_mae: 15.3449 - 38ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 34.1157 - mae: 4.1839 - val_loss: 48.7870 - val_mae: 5.3874 - 40ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 29.8671 - mae: 4.1590 - val_loss: 140.2644 - val_mae: 10.7729 - 35ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 34.7937 - mae: 4.2349 - val_loss: 22.6635 - val_mae: 3.7759 - 40ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 31.2634 - mae: 4.2565 - val_loss: 22.6778 - val_mae: 3.6812 - 41ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 31.7500 - mae: 3.9914 - val_loss: 36.3721 - val_mae: 4.6470 - 38ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 32.1572 - mae: 4.2261 - val_loss: 41.8000 - val_mae: 5.2998 - 39ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 29.4366 - mae: 4.1863 - val_loss: 3

'model_num: 2, params: (13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 38.82, test_mae: 4.84, test_r2: 0.6066'

# Модель 3

In [307]:
model_num = 3

params = '13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = adam, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='relu')(inputs)
outputs = Dense(1)(hidden_layer_1)

model = Model(inputs, outputs)

model.compile(optimizer='adam', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 971.8730 - mae: 22.9079 - val_loss: 160.0877 - val_mae: 9.3585 - 337ms/epoch - 10ms/step
Epoch 2/100
35/35 - 0s - loss: 108.2460 - mae: 7.8144 - val_loss: 98.0509 - val_mae: 7.5873 - 42ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 78.2267 - mae: 6.5364 - val_loss: 98.0261 - val_mae: 7.8113 - 40ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 70.7382 - mae: 6.0806 - val_loss: 94.1552 - val_mae: 7.6416 - 40ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 67.9697 - mae: 5.9205 - val_loss: 119.7369 - val_mae: 9.4733 - 37ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 64.4088 - mae: 5.9157 - val_loss: 72.5267 - val_mae: 5.9293 - 38ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 60.4722 - mae: 5.5194 - val_loss: 74.0712 - val_mae: 6.6358 - 42ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 57.3418 - mae: 5.3308 - val_loss: 70.3303 - val_mae: 6.5171 - 40ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 55.6454 - mae: 5.4363 - val_loss: 77.5929 

Epoch 72/100
35/35 - 0s - loss: 28.3564 - mae: 3.8763 - val_loss: 55.8810 - val_mae: 6.4614 - 40ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 28.1740 - mae: 3.9433 - val_loss: 86.6626 - val_mae: 8.0749 - 39ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 78.2820 - mae: 6.8738 - val_loss: 38.5513 - val_mae: 4.6894 - 37ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 26.9546 - mae: 3.7566 - val_loss: 35.3052 - val_mae: 4.2066 - 41ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 28.7534 - mae: 3.7915 - val_loss: 32.2325 - val_mae: 4.3239 - 40ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 24.3898 - mae: 3.6154 - val_loss: 45.0781 - val_mae: 5.0385 - 40ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 27.7139 - mae: 3.8672 - val_loss: 26.8483 - val_mae: 3.9608 - 39ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 25.2962 - mae: 3.6372 - val_loss: 32.2235 - val_mae: 4.3958 - 39ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 23.1857 - mae: 3.5503 - val_loss: 27.817

'model_num: 3, params: (13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = adam, loss = mse), test_mse: 41.85, test_mae: 4.41, test_r2: 0.5758'

# Модель 4

In [308]:
model_num = 4

params = '13 input, 2 hidden layers (100, 10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='relu')(inputs)
hidden_layer_2 = Dense(10, activation='relu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 240.4370 - mae: 10.2188 - val_loss: 70.1097 - val_mae: 5.7323 - 327ms/epoch - 9ms/step
Epoch 2/100
35/35 - 0s - loss: 76.0482 - mae: 6.3627 - val_loss: 166.0532 - val_mae: 12.0053 - 42ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 63.7109 - mae: 6.1700 - val_loss: 185.3268 - val_mae: 12.6211 - 41ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 64.5520 - mae: 5.9781 - val_loss: 88.7375 - val_mae: 6.5429 - 39ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 58.0561 - mae: 5.5166 - val_loss: 219.1998 - val_mae: 13.6781 - 41ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 62.7084 - mae: 5.8482 - val_loss: 56.2263 - val_mae: 4.3432 - 39ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 59.9761 - mae: 5.6521 - val_loss: 84.4866 - val_mae: 6.8493 - 40ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 53.3739 - mae: 5.4312 - val_loss: 48.7306 - val_mae: 5.4701 - 39ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 51.2488 - mae: 5.3000 - val_loss: 180.87

Epoch 72/100
35/35 - 0s - loss: 26.1554 - mae: 3.6952 - val_loss: 101.3448 - val_mae: 8.6825 - 40ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 24.2991 - mae: 3.6332 - val_loss: 325.5280 - val_mae: 14.8922 - 41ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 28.3763 - mae: 3.6720 - val_loss: 38.2827 - val_mae: 3.8213 - 38ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 24.7412 - mae: 3.6670 - val_loss: 42.5596 - val_mae: 4.7991 - 41ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 27.8152 - mae: 3.7590 - val_loss: 64.8579 - val_mae: 6.2306 - 38ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 23.0737 - mae: 3.5373 - val_loss: 36.5948 - val_mae: 4.2060 - 42ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 24.1843 - mae: 3.4410 - val_loss: 31.4087 - val_mae: 3.3974 - 37ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 23.6151 - mae: 3.5519 - val_loss: 39.5765 - val_mae: 4.9292 - 39ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 22.8540 - mae: 3.5206 - val_loss: 28.

'model_num: 4, params: (13 input, 2 hidden layers (100, 10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 35.54, test_mae: 4.15, test_r2: 0.6398'

# Модель 5

In [309]:
model_num = 5

params = '13 input, 2 hidden layers (100, 10), act = leaky_relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='leaky_relu')(inputs)
hidden_layer_2 = Dense(10, activation='leaky_relu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 273.9569 - mae: 11.4731 - val_loss: 245.1167 - val_mae: 14.0222 - 328ms/epoch - 9ms/step
Epoch 2/100
35/35 - 0s - loss: 118.9019 - mae: 8.2210 - val_loss: 315.5795 - val_mae: 16.2082 - 43ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 105.7436 - mae: 7.8470 - val_loss: 79.3665 - val_mae: 7.7341 - 42ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 83.3840 - mae: 7.1090 - val_loss: 185.4743 - val_mae: 11.6676 - 42ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 94.8677 - mae: 7.3314 - val_loss: 742.5015 - val_mae: 25.3458 - 42ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 107.5291 - mae: 7.8476 - val_loss: 53.6931 - val_mae: 5.8005 - 39ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 83.9271 - mae: 6.9662 - val_loss: 381.2337 - val_mae: 18.0558 - 42ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 77.7113 - mae: 6.7703 - val_loss: 94.9557 - val_mae: 7.5960 - 40ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 83.7228 - mae: 7.0370 - val_loss:

Epoch 72/100
35/35 - 0s - loss: 35.0303 - mae: 4.2876 - val_loss: 108.2726 - val_mae: 9.2205 - 39ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 34.3646 - mae: 4.3676 - val_loss: 413.8893 - val_mae: 18.6393 - 41ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 38.8356 - mae: 4.4143 - val_loss: 52.1519 - val_mae: 5.2335 - 40ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 33.3577 - mae: 4.2735 - val_loss: 73.5640 - val_mae: 7.3104 - 41ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 37.4919 - mae: 4.5058 - val_loss: 76.6394 - val_mae: 7.4555 - 42ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 32.9850 - mae: 4.2527 - val_loss: 26.6529 - val_mae: 3.6996 - 39ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 34.0941 - mae: 4.2924 - val_loss: 23.7646 - val_mae: 3.3747 - 40ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 32.9953 - mae: 4.2743 - val_loss: 54.0430 - val_mae: 5.9484 - 41ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 32.0976 - mae: 4.2091 - val_loss: 56.

'model_num: 5, params: (13 input, 2 hidden layers (100, 10), act = leaky_relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 34.66, test_mae: 4.32, test_r2: 0.6488'

# Модель 6

In [310]:
model_num = 6

params = '13 input, 2 hidden layers (100, 10), act = elu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='elu')(inputs)
hidden_layer_2 = Dense(10, activation='elu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 2125.7830 - mae: 32.8325 - val_loss: 116.2391 - val_mae: 7.8083 - 362ms/epoch - 10ms/step
Epoch 2/100
35/35 - 0s - loss: 82.8528 - mae: 6.3378 - val_loss: 78.5432 - val_mae: 6.9782 - 43ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 61.5015 - mae: 5.7542 - val_loss: 111.4350 - val_mae: 9.4677 - 42ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 60.7383 - mae: 5.6951 - val_loss: 67.1961 - val_mae: 5.1611 - 40ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 56.0077 - mae: 5.2618 - val_loss: 233.9373 - val_mae: 14.1911 - 47ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 59.3441 - mae: 5.6273 - val_loss: 56.7149 - val_mae: 4.5993 - 43ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 55.3623 - mae: 5.3080 - val_loss: 76.1603 - val_mae: 5.9681 - 42ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 49.1063 - mae: 5.1093 - val_loss: 49.5426 - val_mae: 5.1566 - 42ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 47.9768 - mae: 5.0783 - val_loss: 156.83

Epoch 72/100
35/35 - 0s - loss: 26.7076 - mae: 3.7471 - val_loss: 59.0787 - val_mae: 6.5932 - 41ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 25.8097 - mae: 3.6698 - val_loss: 128.6395 - val_mae: 10.0109 - 38ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 28.5482 - mae: 3.8867 - val_loss: 42.3447 - val_mae: 4.9744 - 37ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 26.1566 - mae: 3.7300 - val_loss: 57.4202 - val_mae: 6.2122 - 41ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 27.7176 - mae: 3.6958 - val_loss: 38.1411 - val_mae: 5.2309 - 38ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 25.5194 - mae: 3.6422 - val_loss: 25.0164 - val_mae: 3.9186 - 41ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 25.1325 - mae: 3.5445 - val_loss: 33.0205 - val_mae: 4.1374 - 38ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 25.3642 - mae: 3.6341 - val_loss: 60.4664 - val_mae: 6.6275 - 40ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 26.0350 - mae: 3.7449 - val_loss: 34.1

'model_num: 6, params: (13 input, 2 hidden layers (100, 10), act = elu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 34.48, test_mae: 3.95, test_r2: 0.6506'

# Модель 7

In [311]:
model_num = 7

params = '13 input, 3 hidden layers (30, 20, 10), act = elu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(30, activation='elu')(inputs)
hidden_layer_2 = Dense(20, activation='elu')(hidden_layer_1)
hidden_layer_3 = Dense(10, activation='elu')(hidden_layer_2)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=100,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/100
35/35 - 0s - loss: 550.9799 - mae: 16.8676 - val_loss: 64.7453 - val_mae: 5.5511 - 339ms/epoch - 10ms/step
Epoch 2/100
35/35 - 0s - loss: 79.2064 - mae: 6.6699 - val_loss: 253.0543 - val_mae: 14.6530 - 42ms/epoch - 1ms/step
Epoch 3/100
35/35 - 0s - loss: 71.1905 - mae: 6.5399 - val_loss: 132.2360 - val_mae: 10.8390 - 42ms/epoch - 1ms/step
Epoch 4/100
35/35 - 0s - loss: 67.7798 - mae: 6.1205 - val_loss: 86.8939 - val_mae: 6.6551 - 43ms/epoch - 1ms/step
Epoch 5/100
35/35 - 0s - loss: 64.9366 - mae: 5.8755 - val_loss: 284.9841 - val_mae: 15.5971 - 40ms/epoch - 1ms/step
Epoch 6/100
35/35 - 0s - loss: 69.5632 - mae: 6.2010 - val_loss: 46.9698 - val_mae: 3.9078 - 42ms/epoch - 1ms/step
Epoch 7/100
35/35 - 0s - loss: 60.2864 - mae: 5.7360 - val_loss: 155.4656 - val_mae: 10.1946 - 42ms/epoch - 1ms/step
Epoch 8/100
35/35 - 0s - loss: 56.4438 - mae: 5.5474 - val_loss: 49.2896 - val_mae: 5.5162 - 40ms/epoch - 1ms/step
Epoch 9/100
35/35 - 0s - loss: 58.1201 - mae: 5.6902 - val_loss: 101

Epoch 72/100
35/35 - 0s - loss: 35.2061 - mae: 4.4599 - val_loss: 95.4013 - val_mae: 8.6472 - 38ms/epoch - 1ms/step
Epoch 73/100
35/35 - 0s - loss: 32.9150 - mae: 4.3125 - val_loss: 230.2586 - val_mae: 13.0464 - 39ms/epoch - 1ms/step
Epoch 74/100
35/35 - 0s - loss: 35.9855 - mae: 4.3588 - val_loss: 46.8076 - val_mae: 5.1171 - 37ms/epoch - 1ms/step
Epoch 75/100
35/35 - 0s - loss: 33.8615 - mae: 4.3594 - val_loss: 135.6468 - val_mae: 10.4135 - 38ms/epoch - 1ms/step
Epoch 76/100
35/35 - 0s - loss: 36.6592 - mae: 4.3548 - val_loss: 36.2439 - val_mae: 5.0705 - 40ms/epoch - 1ms/step
Epoch 77/100
35/35 - 0s - loss: 33.5988 - mae: 4.3442 - val_loss: 26.1759 - val_mae: 4.0085 - 37ms/epoch - 1ms/step
Epoch 78/100
35/35 - 0s - loss: 34.4684 - mae: 4.2025 - val_loss: 35.7330 - val_mae: 4.2860 - 40ms/epoch - 1ms/step
Epoch 79/100
35/35 - 0s - loss: 34.9766 - mae: 4.3989 - val_loss: 57.7726 - val_mae: 6.5523 - 39ms/epoch - 1ms/step
Epoch 80/100
35/35 - 0s - loss: 32.7429 - mae: 4.4267 - val_loss: 28

'model_num: 7, params: (13 input, 3 hidden layers (30, 20, 10), act = elu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse), test_mse: 36.51, test_mae: 4.35, test_r2: 0.63'

# Модель 8

In [312]:
model_num = 8

params = '13 input, 2 hidden layers (100, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='elu')(inputs)
hidden_layer_2 = Dense(10, activation='elu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=200,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/200
35/35 - 0s - loss: 135.1587 - mae: 8.4342 - val_loss: 72.5591 - val_mae: 5.9180 - 339ms/epoch - 10ms/step
Epoch 2/200
35/35 - 0s - loss: 80.7531 - mae: 6.6309 - val_loss: 85.4300 - val_mae: 6.6972 - 42ms/epoch - 1ms/step
Epoch 3/200
35/35 - 0s - loss: 73.4186 - mae: 6.4026 - val_loss: 147.1578 - val_mae: 11.3108 - 41ms/epoch - 1ms/step
Epoch 4/200
35/35 - 0s - loss: 66.9427 - mae: 6.1380 - val_loss: 86.9284 - val_mae: 6.3259 - 42ms/epoch - 1ms/step
Epoch 5/200
35/35 - 0s - loss: 64.3676 - mae: 5.9856 - val_loss: 149.6914 - val_mae: 10.9829 - 41ms/epoch - 1ms/step
Epoch 6/200
35/35 - 0s - loss: 65.4001 - mae: 6.1592 - val_loss: 51.5608 - val_mae: 4.7834 - 41ms/epoch - 1ms/step
Epoch 7/200
35/35 - 0s - loss: 59.8597 - mae: 5.7408 - val_loss: 101.3051 - val_mae: 7.2453 - 40ms/epoch - 1ms/step
Epoch 8/200
35/35 - 0s - loss: 58.1166 - mae: 5.7406 - val_loss: 103.2980 - val_mae: 8.0371 - 41ms/epoch - 1ms/step
Epoch 9/200
35/35 - 0s - loss: 55.4047 - mae: 5.5685 - val_loss: 133.42

Epoch 72/200
35/35 - 0s - loss: 25.9704 - mae: 3.5526 - val_loss: 100.2006 - val_mae: 8.4359 - 41ms/epoch - 1ms/step
Epoch 73/200
35/35 - 0s - loss: 23.2804 - mae: 3.5075 - val_loss: 188.6672 - val_mae: 11.3334 - 39ms/epoch - 1ms/step
Epoch 74/200
35/35 - 0s - loss: 24.8547 - mae: 3.4892 - val_loss: 64.7021 - val_mae: 6.6296 - 41ms/epoch - 1ms/step
Epoch 75/200
35/35 - 0s - loss: 21.7009 - mae: 3.3786 - val_loss: 28.2778 - val_mae: 3.7110 - 41ms/epoch - 1ms/step
Epoch 76/200
35/35 - 0s - loss: 24.5546 - mae: 3.5538 - val_loss: 25.7582 - val_mae: 4.1447 - 42ms/epoch - 1ms/step
Epoch 77/200
35/35 - 0s - loss: 21.5089 - mae: 3.4483 - val_loss: 23.8332 - val_mae: 3.3223 - 41ms/epoch - 1ms/step
Epoch 78/200
35/35 - 0s - loss: 22.2192 - mae: 3.3381 - val_loss: 45.8413 - val_mae: 5.4477 - 42ms/epoch - 1ms/step
Epoch 79/200
35/35 - 0s - loss: 22.4368 - mae: 3.5053 - val_loss: 54.4522 - val_mae: 5.9177 - 38ms/epoch - 1ms/step
Epoch 80/200
35/35 - 0s - loss: 21.0451 - mae: 3.4042 - val_loss: 35.

Epoch 143/200
35/35 - 0s - loss: 16.2940 - mae: 2.9908 - val_loss: 23.8778 - val_mae: 3.1470 - 40ms/epoch - 1ms/step
Epoch 144/200
35/35 - 0s - loss: 17.6335 - mae: 3.0284 - val_loss: 21.3146 - val_mae: 3.2810 - 40ms/epoch - 1ms/step
Epoch 145/200
35/35 - 0s - loss: 16.8651 - mae: 3.0861 - val_loss: 22.0390 - val_mae: 3.3307 - 41ms/epoch - 1ms/step
Epoch 146/200
35/35 - 0s - loss: 16.9500 - mae: 2.9902 - val_loss: 27.7957 - val_mae: 4.1213 - 39ms/epoch - 1ms/step
Epoch 147/200
35/35 - 0s - loss: 18.1387 - mae: 3.1582 - val_loss: 31.5117 - val_mae: 3.9886 - 42ms/epoch - 1ms/step
Epoch 148/200
35/35 - 0s - loss: 17.4362 - mae: 3.0399 - val_loss: 40.5261 - val_mae: 5.0781 - 41ms/epoch - 1ms/step
Epoch 149/200
35/35 - 0s - loss: 17.2910 - mae: 3.0627 - val_loss: 23.6045 - val_mae: 3.5460 - 45ms/epoch - 1ms/step
Epoch 150/200
35/35 - 0s - loss: 15.4173 - mae: 2.7976 - val_loss: 26.1122 - val_mae: 3.6081 - 42ms/epoch - 1ms/step
Epoch 151/200
35/35 - 0s - loss: 17.7973 - mae: 3.0352 - val_los

'model_num: 8, params: (13 input, 2 hidden layers (100, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse), test_mse: 20.31, test_mae: 3.18, test_r2: 0.7942'

# Модель 9

In [313]:
model_num = 9

params = '13 input, 2 hidden layers (100, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(100, activation='elu')(inputs)
hidden_layer_2 = Dense(10, activation='elu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=200,
                    batch_size = 50,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/200
7/7 - 0s - loss: 119.5015 - mae: 8.3443 - val_loss: 145.7955 - val_mae: 11.0634 - 314ms/epoch - 45ms/step
Epoch 2/200
7/7 - 0s - loss: 67.2962 - mae: 6.3425 - val_loss: 69.9517 - val_mae: 5.7695 - 24ms/epoch - 3ms/step
Epoch 3/200
7/7 - 0s - loss: 61.6657 - mae: 5.6235 - val_loss: 65.3017 - val_mae: 5.9186 - 22ms/epoch - 3ms/step
Epoch 4/200
7/7 - 0s - loss: 67.0725 - mae: 6.0669 - val_loss: 78.9852 - val_mae: 5.9980 - 24ms/epoch - 3ms/step
Epoch 5/200
7/7 - 0s - loss: 59.9856 - mae: 5.5649 - val_loss: 65.0487 - val_mae: 6.2370 - 22ms/epoch - 3ms/step
Epoch 6/200
7/7 - 0s - loss: 61.9143 - mae: 5.7556 - val_loss: 77.1531 - val_mae: 5.8748 - 24ms/epoch - 3ms/step
Epoch 7/200
7/7 - 0s - loss: 65.3260 - mae: 5.8357 - val_loss: 72.0789 - val_mae: 7.3899 - 23ms/epoch - 3ms/step
Epoch 8/200
7/7 - 0s - loss: 56.3615 - mae: 5.3879 - val_loss: 65.2394 - val_mae: 6.8687 - 23ms/epoch - 3ms/step
Epoch 9/200
7/7 - 0s - loss: 68.2399 - mae: 6.2546 - val_loss: 58.3312 - val_mae: 5.3020 - 

Epoch 73/200
7/7 - 0s - loss: 36.3681 - mae: 4.4479 - val_loss: 30.7551 - val_mae: 3.9241 - 27ms/epoch - 4ms/step
Epoch 74/200
7/7 - 0s - loss: 28.8259 - mae: 3.8241 - val_loss: 31.6498 - val_mae: 4.0043 - 23ms/epoch - 3ms/step
Epoch 75/200
7/7 - 0s - loss: 46.3136 - mae: 5.1043 - val_loss: 31.4207 - val_mae: 3.9511 - 24ms/epoch - 3ms/step
Epoch 76/200
7/7 - 0s - loss: 32.9025 - mae: 4.0173 - val_loss: 40.8947 - val_mae: 5.0836 - 22ms/epoch - 3ms/step
Epoch 77/200
7/7 - 0s - loss: 37.3472 - mae: 4.3768 - val_loss: 36.4282 - val_mae: 4.7353 - 25ms/epoch - 4ms/step
Epoch 78/200
7/7 - 0s - loss: 29.4066 - mae: 3.8232 - val_loss: 39.0312 - val_mae: 4.8749 - 22ms/epoch - 3ms/step
Epoch 79/200
7/7 - 0s - loss: 43.1057 - mae: 5.1044 - val_loss: 33.7206 - val_mae: 4.1565 - 24ms/epoch - 3ms/step
Epoch 80/200
7/7 - 0s - loss: 33.0624 - mae: 4.1383 - val_loss: 52.8776 - val_mae: 6.1829 - 21ms/epoch - 3ms/step
Epoch 81/200
7/7 - 0s - loss: 31.3430 - mae: 4.0413 - val_loss: 38.7930 - val_mae: 4.267

Epoch 145/200
7/7 - 0s - loss: 27.6656 - mae: 3.8131 - val_loss: 53.7950 - val_mae: 6.0495 - 24ms/epoch - 3ms/step
Epoch 146/200
7/7 - 0s - loss: 30.7855 - mae: 4.1165 - val_loss: 28.6161 - val_mae: 4.0699 - 22ms/epoch - 3ms/step
Epoch 147/200
7/7 - 0s - loss: 22.7977 - mae: 3.4510 - val_loss: 31.8090 - val_mae: 4.3548 - 24ms/epoch - 3ms/step
Epoch 148/200
7/7 - 0s - loss: 29.9986 - mae: 4.2005 - val_loss: 50.9945 - val_mae: 5.9421 - 23ms/epoch - 3ms/step
Epoch 149/200
7/7 - 0s - loss: 25.6339 - mae: 3.8188 - val_loss: 58.6072 - val_mae: 6.2205 - 24ms/epoch - 3ms/step
Epoch 150/200
7/7 - 0s - loss: 30.9028 - mae: 4.2482 - val_loss: 39.5196 - val_mae: 4.5539 - 22ms/epoch - 3ms/step
Epoch 151/200
7/7 - 0s - loss: 26.3875 - mae: 3.8051 - val_loss: 46.9294 - val_mae: 5.5970 - 23ms/epoch - 3ms/step
Epoch 152/200
7/7 - 0s - loss: 26.5470 - mae: 3.8215 - val_loss: 33.8566 - val_mae: 4.3981 - 21ms/epoch - 3ms/step
Epoch 153/200
7/7 - 0s - loss: 34.1793 - mae: 4.3253 - val_loss: 37.3515 - val_m

'model_num: 9, params: (13 input, 2 hidden layers (100, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse), test_mse: 60.52, test_mae: 5.5, test_r2: 0.3867'

# Модель 10

In [314]:
model_num = 10

params = '13 input, 2 hidden layers (50, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse'

inputs = Input(shape=(13))
hidden_layer_1 = Dense(50, activation='elu')(inputs)
hidden_layer_2 = Dense(10, activation='elu')(hidden_layer_1)
outputs = Dense(1)(hidden_layer_2)

model = Model(inputs, outputs)

model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])

history = model.fit(x_train, 
                    y_train, 
                    epochs=200,
                    batch_size = 10,
                    validation_split=0.1, 
                    verbose=2)


y_test_pred = model.predict(x_test)
test_mse, test_mae = model.evaluate(x_test, y_test, verbose=2)

test_r2 = r2_score(y_test, y_test_pred)

result_list.append([model_num, params, test_mse, test_mae, test_r2])
print 
(f'model_num: {model_num}, params: ({params}), test_mse: {round(test_mse, 2)}, test_mae: {round(test_mae,2)}, test_r2: {round(test_r2,4)}')

Epoch 1/200
35/35 - 0s - loss: 265.9599 - mae: 11.0261 - val_loss: 166.4753 - val_mae: 10.8958 - 337ms/epoch - 10ms/step
Epoch 2/200
35/35 - 0s - loss: 113.3240 - mae: 8.2039 - val_loss: 50.4806 - val_mae: 5.2340 - 41ms/epoch - 1ms/step
Epoch 3/200
35/35 - 0s - loss: 88.6513 - mae: 7.5185 - val_loss: 164.4128 - val_mae: 11.6338 - 43ms/epoch - 1ms/step
Epoch 4/200
35/35 - 0s - loss: 87.1717 - mae: 7.0458 - val_loss: 150.5339 - val_mae: 10.0022 - 39ms/epoch - 1ms/step
Epoch 5/200
35/35 - 0s - loss: 85.4409 - mae: 6.9487 - val_loss: 398.9881 - val_mae: 18.4365 - 39ms/epoch - 1ms/step
Epoch 6/200
35/35 - 0s - loss: 84.8710 - mae: 6.9611 - val_loss: 49.3995 - val_mae: 5.1227 - 42ms/epoch - 1ms/step
Epoch 7/200
35/35 - 0s - loss: 76.5036 - mae: 6.5885 - val_loss: 181.5706 - val_mae: 11.6220 - 41ms/epoch - 1ms/step
Epoch 8/200
35/35 - 0s - loss: 68.4507 - mae: 6.3061 - val_loss: 130.3450 - val_mae: 10.6455 - 44ms/epoch - 1ms/step
Epoch 9/200
35/35 - 0s - loss: 65.5961 - mae: 6.1692 - val_loss

Epoch 72/200
35/35 - 0s - loss: 36.5739 - mae: 4.6140 - val_loss: 97.2379 - val_mae: 8.5614 - 39ms/epoch - 1ms/step
Epoch 73/200
35/35 - 0s - loss: 37.3657 - mae: 4.5255 - val_loss: 400.9080 - val_mae: 17.6151 - 39ms/epoch - 1ms/step
Epoch 74/200
35/35 - 0s - loss: 41.0444 - mae: 4.4974 - val_loss: 59.6379 - val_mae: 5.7436 - 40ms/epoch - 1ms/step
Epoch 75/200
35/35 - 0s - loss: 35.2271 - mae: 4.5583 - val_loss: 156.6895 - val_mae: 11.3706 - 38ms/epoch - 1ms/step
Epoch 76/200
35/35 - 0s - loss: 41.2925 - mae: 4.7571 - val_loss: 29.7762 - val_mae: 4.4441 - 38ms/epoch - 1ms/step
Epoch 77/200
35/35 - 0s - loss: 34.5435 - mae: 4.3598 - val_loss: 27.4726 - val_mae: 4.1214 - 40ms/epoch - 1ms/step
Epoch 78/200
35/35 - 0s - loss: 35.7222 - mae: 4.2978 - val_loss: 27.5420 - val_mae: 3.8615 - 41ms/epoch - 1ms/step
Epoch 79/200
35/35 - 0s - loss: 36.6936 - mae: 4.5279 - val_loss: 46.9133 - val_mae: 5.8747 - 38ms/epoch - 1ms/step
Epoch 80/200
35/35 - 0s - loss: 34.8155 - mae: 4.5220 - val_loss: 34

Epoch 143/200
35/35 - 0s - loss: 21.3460 - mae: 3.4535 - val_loss: 56.6992 - val_mae: 5.5475 - 45ms/epoch - 1ms/step
Epoch 144/200
35/35 - 0s - loss: 21.1627 - mae: 3.4898 - val_loss: 37.6336 - val_mae: 4.6581 - 45ms/epoch - 1ms/step
Epoch 145/200
35/35 - 0s - loss: 19.6790 - mae: 3.2500 - val_loss: 21.2333 - val_mae: 3.1331 - 42ms/epoch - 1ms/step
Epoch 146/200
35/35 - 0s - loss: 19.8345 - mae: 3.3338 - val_loss: 25.9253 - val_mae: 3.3616 - 44ms/epoch - 1ms/step
Epoch 147/200
35/35 - 0s - loss: 22.7102 - mae: 3.4969 - val_loss: 30.4542 - val_mae: 3.7067 - 45ms/epoch - 1ms/step
Epoch 148/200
35/35 - 0s - loss: 20.5216 - mae: 3.1977 - val_loss: 31.4550 - val_mae: 3.5312 - 44ms/epoch - 1ms/step
Epoch 149/200
35/35 - 0s - loss: 20.5322 - mae: 3.3704 - val_loss: 21.4531 - val_mae: 2.9312 - 42ms/epoch - 1ms/step
Epoch 150/200
35/35 - 0s - loss: 19.1053 - mae: 3.2776 - val_loss: 49.2661 - val_mae: 4.9088 - 47ms/epoch - 1ms/step
Epoch 151/200
35/35 - 0s - loss: 20.1899 - mae: 3.2659 - val_los

'model_num: 10, params: (13 input, 2 hidden layers (50, 10), act = elu, epochs = 200, batch size = 10, opt = rmsprop, loss = mse), test_mse: 23.11, test_mae: 3.35, test_r2: 0.7658'

In [315]:
result_list

[[1,
  '13 input, 1 hidden layer (10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse',
  36.2992057800293,
  4.2866740226745605,
  0.6321279378809561],
 [2,
  '13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse',
  38.82135772705078,
  4.843014717102051,
  0.6065673021779683],
 [3,
  '13 input, 1 hidden layer (100), act = relu, epochs = 100, batch size = 10, opt = adam, loss = mse',
  41.85401916503906,
  4.407288074493408,
  0.5758330168402424],
 [4,
  '13 input, 2 hidden layers (100, 10), act = relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse',
  35.54302978515625,
  4.149533748626709,
  0.639791309507479],
 [5,
  '13 input, 2 hidden layers (100, 10), act = leaky_relu, epochs = 100, batch size = 10, opt = rmsprop, loss = mse',
  34.65501022338867,
  4.320472240447998,
  0.6487908678623907],
 [6,
  '13 input, 2 hidden layers (100, 10), act = elu, epochs = 100, batch size = 10, opt = rmsprop, loss = 

In [323]:
# создадим датасет-таблицу для удобства интерпретации результатов
result_df = pd.DataFrame(result_list, columns = ['model_num', 'params', 'test_mse', 'test_mae', 'test_r2']).round(4)

In [324]:
# отсортируем данные таблице по показателю r2
result_df.sort_values(['test_r2'], ascending = False)

model_num                                             params  test_mse  \
7          8  13 input, 2 hidden layers (100, 10), act = elu...   20.3061   
9         10  13 input, 2 hidden layers (50, 10), act = elu,...   23.1132   
5          6  13 input, 2 hidden layers (100, 10), act = elu...   34.4754   
4          5  13 input, 2 hidden layers (100, 10), act = lea...   34.6550   
3          4  13 input, 2 hidden layers (100, 10), act = rel...   35.5430   
0          1  13 input, 1 hidden layer (10), act = relu, epo...   36.2992   
6          7  13 input, 3 hidden layers (30, 20, 10), act = ...   36.5106   
1          2  13 input, 1 hidden layer (100), act = relu, ep...   38.8214   
2          3  13 input, 1 hidden layer (100), act = relu, ep...   41.8540   
8          9  13 input, 2 hidden layers (100, 10), act = elu...   60.5167   

   test_mae  test_r2  
7    3.1799   0.7942  
9    3.3549   0.7658  
5    3.9526   0.6506  
4    4.3205   0.6488  
3    4.1495   0.6398  
0    4.2867   0.6321  
6    4.3531   0.6300  
1    4.8430   0.6066  
2    4.4073   0.5758  
8    5.4954   0.3867

### Вывод: лучшую метрику r2 для задачи регрессии показала модель 8 с параметрами: 2 hidden layers (100, 10 neurons), activation elu, 200 epochs, batch size 10, optimizer = rmsprop, loss mse. При таких параметрах r2 составляет 0.7942